File: fix_encoding.py

import os
import chardet

def detect_encoding(file_path):
    with open(file_path, 'rb') as f:
        raw_data = f.read()
        result = chardet.detect(raw_data)
        return result['encoding'], result['confidence']

def fix_file_encoding(file_path):
    try:
        with open(file_path, 'rb') as f:
            raw_data = f.read()
        
        # 尝试多种编码方案
        encodings_to_try = ['utf-8', 'gbk', 'gb2312', 'big5', 'iso-8859-1']
        content = None
        used_encoding = None
        
        for encoding in encodings_to_try:
            try:
                content = raw_data.decode(encoding)
                used_encoding = encoding
                print(f"  成功使用 {encoding} 解码")
                break
            except UnicodeDecodeError:
                continue
        
        if content is None:
            print(f"  无法解码文件: {file_path}")
            return False
        
        # 检查是否需要修复（常见的UTF-8被错误保存为GBK的情况）
        # 尝试将内容作为UTF-8重新编码，看看是否能得到合理的中文
        try:
            # 尝试修复：内容看起来像是GBK编码被错误解释为ISO-8859-1
            test_content = content.encode('latin-1').decode('utf-8')
            # 检查是否包含有效中文字符
            has_chinese = any('\u4e00' &lt;= c &lt;= '\u9fff' for c in test_content)
            if has_chinese:
                content = test_content
                print(f"  发现并修复了编码转换问题")
        except:
            pass
        
        # 保存为UTF-8
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(content)
        
        print(f"  已修复: {file_path}")
        return True
        
    except Exception as e:
        print(f"  处理文件失败 {file_path}: {e}")
        return False

def main():
    root_dir = '.'
    extensions = ['.md', '.java', '.xml', '.properties']
    
    print(f"开始扫描目录: {os.path.abspath(root_dir)}")
    print(f"处理文件类型: {extensions}")
    
    fixed_count = 0
    total_count = 0
    
    for dirpath, dirnames, filenames in os.walk(root_dir):
        # 跳过.git目录
        if '.git' in dirpath:
            continue
            
        for filename in filenames:
            ext = os.path.splitext(filename)[1].lower()
            if ext in extensions:
                file_path = os.path.join(dirpath, filename)
                total_count += 1
                print(f"\n处理文件 {total_count}: {file_path}")
                if fix_file_encoding(file_path):
                    fixed_count += 1
    
    print(f"\n完成！共处理 {total_count} 个文件，修复了 {fixed_count} 个文件。")

if __name__ == '__main__':
    main()

File: FixEncoding.java

import java.io.*;
import java.nio.charset.*;
import java.nio.file.*;
import java.util.*;

public class FixEncoding {

    private static final List&lt;String&gt; SUPPORTED_EXTENSIONS = Arrays.asList(".md", ".java", ".xml", ".properties");
    private static final Charset TARGET_CHARSET = StandardCharsets.UTF_8;

    public static void main(String[] args) {
        String rootDir = args.length &gt; 0 ? args[0] : System.getProperty("user.dir");
        System.out.println("Start fixing file encoding, root dir: " + rootDir);

        try {
            Files.walk(Paths.get(rootDir))
                .filter(Files::isRegularFile)
                .filter(path -&gt; {
                    String fileName = path.toString().toLowerCase();
                    return SUPPORTED_EXTENSIONS.stream().anyMatch(fileName::endsWith);
                })
                .filter(path -&gt; !path.toString().contains(".git"))
                .forEach(FixEncoding::fixFileEncoding);

            System.out.println("Encoding fix complete!");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void fixFileEncoding(Path path) {
        try {
            System.out.println("Processing: " + path);
            byte[] bytes = Files.readAllBytes(path);

            String content = null;

            try {
                content = new String(bytes, StandardCharsets.UTF_8);
            } catch (Exception e) {
            }

            if (content != null &amp;&amp; (content.contains("\uFFFD") || content.contains("\u951F"))) {
                System.out.println("  Garbled detected, trying to fix...");

                try {
                    String testContent = new String(content.getBytes(StandardCharsets.ISO_8859_1), StandardCharsets.UTF_8);
                    if (containsValidChinese(testContent)) {
                        content = testContent;
                        System.out.println("  Fixed using method 1 (UTF-8-&gt;ISO-8859-1-&gt;UTF-8)");
                    }
                } catch (Exception e) {
                }

                if (content.contains("\uFFFD") || content.contains("\u951F")) {
                    try {
                        Charset gbkCharset = Charset.forName("GBK");
                        String testContent = new String(bytes, gbkCharset);
                        if (containsValidChinese(testContent) &amp;&amp; !testContent.contains("\uFFFD")) {
                            content = testContent;
                            System.out.println("  Fixed using method 2 (GBK-&gt;UTF-8)");
                        }
                    } catch (Exception e) {
                    }
                }
            }

            if (content != null) {
                Files.write(path, content.getBytes(TARGET_CHARSET));
                System.out.println("  Fixed: " + path);
            }

        } catch (IOException e) {
            System.err.println("Failed to process: " + path);
            e.printStackTrace();
        }
    }

    private static boolean containsValidChinese(String content) {
        int chineseCount = 0;
        for (char c : content.toCharArray()) {
            if (c &gt;= 0x4E00 &amp;&amp; c &lt;= 0x9FFF) {
                chineseCount++;
            }
        }
        return chineseCount &gt;= 5;
    }
}

